cmd/image-autobumper/imagebumper/imagebumper.go

```diff
@@ -0,0 +1,281 @@
+
+// DeconstructCommit separates a git describe commit into its parts.
+
+// Examples:
+//
+//	v0.0.30-14-gdeadbeef => (v0.0.30 14 deadbeef)
+//	v0.0.30 => (v0.0.30 0 "")
+//	deadbeef => ("", 0, deadbeef)
+//
+// See man git describe.
+func DeconstructCommit(commit string) (string, int, string) {
+	parts := commitRegexp.FindStringSubmatch(commit)
+	if parts == nil {
+		return "", 0, ""
+	}
+	if parts[1] != "" {
+		return "", 0, parts[1]
+	}
+	var n int
+	if s := parts[3]; s != "" {
+		var err error
+		n, err = strconv.Atoi(s)
+		if err != nil {
+			panic(err)
+		}
+	}
+	return parts[2], n, parts[4]
+}
+
+// DeconstructTag separates the tag into its vDATE-COMMIT-VARIANT components
+//
+// COMMIT may be in the form vTAG-NEW-gCOMMIT, use PureCommit to further process
+// this down to COMMIT.
+func DeconstructTag(tag string) (date, commit, variant string) {
+	currentTagParts := tagRegexp.FindStringSubmatch(tag)
+	if currentTagParts == nil {
+		return "", "", ""
+	}
+	parts := strings.Split(currentTagParts[tagVersionPart], "-")
+	return parts[0][1:], parts[len(parts)-1], currentTagParts[tagExtraPart]
+}
+
+func (cli *Client) getManifest(imageHost, imageName string) (manifest, error) {
+	resp, err := cli.httpClient.Get("https://" + imageHost + "/v2/" + imageName + "/tags/list")
+	if err != nil {
+		return nil, fmt.Errorf("couldn't fetch tag list: %w", err)
+	}
+	defer resp.Body.Close()
+
+	result := struct {
+		Manifest manifest `json:"manifest"`
+	}{}
+
+	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
+		return nil, fmt.Errorf("couldn't parse tag information from registry: %w", err)
+	}
+
+	return result.Manifest, nil
+}
+
+// FindLatestTag returns the latest valid tag for the given image.
+func (cli *Client) FindLatestTag(imageHost, imageName, currentTag string) (string, error) {
+	k := imageHost + "/" + imageName + ":" + currentTag
+	if result, ok := cli.tagCache[k]; ok {
+		return result, nil
+	}
+
+	currentTagParts := tagRegexp.FindStringSubmatch(currentTag)
+	if currentTagParts == nil {
+		return "", fmt.Errorf("couldn't figure out the current tag in %q", currentTag)
+	}
+	if currentTagParts[tagVersionPart] == "latest" {
+		return currentTag, nil
+	}
+
+	imageList, err := cli.getManifest(imageHost, imageName)
+	if err != nil {
+		return "", err
+	}
+
+	latestTag, err := pickBestTag(currentTagParts, imageList)
+	if err != nil {
+		return "", err
+	}
+
+	cli.tagCache[k] = latestTag
+
+	return latestTag, nil
+}
+
+func (cli *Client) TagExists(imageHost, imageName, currentTag string) (bool, error) {
+	imageList, err := cli.getManifest(imageHost, imageName)
+	if err != nil {
+		return false, err
+	}
+
+	for _, v := range imageList {
+		for _, tag := range v.Tags {
+			if tag == currentTag {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func pickBestTag(currentTagParts []string, manifest manifest) (string, error) {
+	// The approach is to find the most recently created image that has the same suffix as the
```

Comment:
Either comment or unit test could be useful to explain what this is about.

Reply:
Tests and comment added